pkg/controller/volume/snapshot/snapshot_controller.go

```diff
@@ -0,0 +1,440 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package volume implements a controller to manage snapshot operations.
+package snapshot
+
+import (
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	"k8s.io/kubernetes/pkg/client/record"
+	"k8s.io/kubernetes/pkg/cloudprovider"
+	"k8s.io/kubernetes/pkg/controller/framework"
+	"k8s.io/kubernetes/pkg/controller/volume/snapshot/cache"
+	"k8s.io/kubernetes/pkg/controller/volume/snapshot/reconciler"
+	"k8s.io/kubernetes/pkg/types"
+	"k8s.io/kubernetes/pkg/util/io"
+	"k8s.io/kubernetes/pkg/util/mount"
+	"k8s.io/kubernetes/pkg/util/runtime"
+	"k8s.io/kubernetes/pkg/volume"
+	"k8s.io/kubernetes/pkg/volume/util/operationexecutor"
+	"k8s.io/kubernetes/pkg/volume/util/volumehelper"
+)
+
+const (
+	// loopPeriod is the amount of time the reconciler loop waits between
+	// successive executions
+	reconcilerLoopPeriod time.Duration = 100 * time.Millisecond
+)
+
+// SnapshotController defines the operations supported by this controller.
+type SnapshotController interface {
+	Run(stopCh <-chan struct{})
+}
+
+// NewSnapshotController returns a new instance of SnapshotController.
+func NewSnapshotController(
+	kubeClient internalclientset.Interface,
+	recorder record.EventRecorder,
+	pvcInformer framework.SharedInformer,
+	pvInformer framework.SharedInformer,
+	cloud cloudprovider.Interface,
+	plugins []volume.VolumePlugin) (SnapshotController, error) {
+	sc := &snapshotController{
+		kubeClient:  kubeClient,
+		pvcInformer: pvcInformer,
+		pvInformer:  pvInformer,
+		cloud:       cloud,
+	}
+
+	sc.pvcInformer.AddEventHandler(framework.ResourceEventHandlerFuncs{
+		AddFunc:    sc.pvcAdd,
+		UpdateFunc: sc.pvcUpdate,
+		DeleteFunc: sc.pvcDelete,
+	})
+
+	if err := sc.volumePluginMgr.InitPlugins(plugins, sc); err != nil {
```

Comment:
Ideally it should be called once per binary, but since they are separate instances it won't cause any problems. We should refactor this in the future.